### PR TITLE
Created ivector-extract-dense.cc. Extracts ivectors from sliding window.

### PR DESCRIPTION
**Pull request by mmaciej2**
Current version uses chunk-size and period to extract a series of ivectors from each segment and stores the resulting matrix.

### CODE REVIEW

File: src/ivectorbin/ivector-extract-dense.cc [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      // Handle the final chunk of irregular size
+      if (imperfect_size) {
+        Vector<BaseFloat> ivector(extractor.IvectorDim());
+        SubMatrix<BaseFloat> sub_mat(mat, num_chunks * period, final_chunk_size, 0, mat.NumCols());
+        IvectorExtract(extractor, utt, Matrix<BaseFloat>(sub_mat),
+         	       std::vector<std::vector<std::pair<int32, BaseFloat> > >
+         	       (&posterior[num_chunks * period], &posterior[num_chunks * period + final_chunk_size]),
+        	       &ivector, auxf_ptr);
+        ivectors.CopyRowFromVec(ivector, num_chunks);
+      }
```
Comment (david-ryan-snyder) [inline src/ivectorbin/ivector-extract-dense.cc]: This is pretty good, but I think it will be easier to maintain if you get rid of the if (imperfect_size) { } block, and find a way to include that logic succinctly in the main loop. You can use the `std::min()` function to help you pick a `chunk_size` that fits within the feature matrix. For example, when you're taking the subset of the posteriors, you can just do something like the following on line 77:

`&posterior[i * period], &posterior[std::min(mat.NumRows(),  i * period + chunk_size)]`

You can do something similar to get the feature submatrix. 

Also, the variable `num_chunks`, is really `num_chunks - 1` in your code. This could confuse a reader of your code. It would be better to define it as `int32 num_chunks = (mat.NumRows() - chunk_size + period) / period + 1;` to avoid confusion.

**Comment by mmaciej2** [comment]
I was doing something along those lines originally, but it runs into a undesirable edge case. Unless I'm mistaken, if a chunk ends perfectly in line with the segment, it will erroneously create another ivector from the last (chunk_size - period) frames, which have all already been covered by the previous chunk. I ended up doing what I did to solve that problem.

That being said, I think there is probably a more elegant solution that I'll work out.

**Comment by david-ryan-snyder** [comment]
Yeah, I see what you mean. You should be able to use a combination of the ceiling function and casting one of the ints to a float to handle that. For example, this might work:

`int32 num_chunks = ceil((mat.NumRows() - chunk_size)
            / static_cast<BaseFloat>(period)) + 1;`

I think it's always best to make the variable and function names as self documenting as possible. So if the a variable is called `num_chunks,` a reader should know what it means by just looking at the name.

Anyway,  by using something like the above snippet, you should be able to get rid of that `if (imperfect) {}` block, and handle all that within the loop. This should make it easier for you to extend this program later.

**Comment by david-ryan-snyder** [comment]
If you can't use the ceiling function, see if you need to include one of the following
 `#include "base/kaldi-common.h"` or 
`#include "util/common-utils.h"
`

**Comment by mmaciej2** [comment]
I don't know how I missed it, but the ceiling function was definitely the solution to everything. I ran a couple test runs for the edge cases and it does appear to be handling them properly.
